File: Week01/subset.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Subset returns a subset of backends according to clientId and subsetSize
func Subset(backends []int, clientID int, subsetSize int) []int {
	subsetCount := len(backends) / subsetSize
	round := clientID / subsetCount
	rand.Seed(int64(round))
	rand.Shuffle(len(backends), func(i, j int) {
		backends[i], backends[j] = backends[j], backends[i]
	})
	subsetID := clientID % subsetCount
	start := subsetID % subsetSize
	return backends[start : start+subsetSize]
}

func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	subsetSize := 10
	clientID := rand.Intn(100)
	backends := make([]int, 100)
	for i := range backends {
		backends[i] = rand.Intn(10000)
	}
	fmt.Printf("clientID: %d, subsetSize: %d\n\nbackends list: %v\n\n", clientID, subsetSize, backends)
	for i := 0; i < 10; i++ {
		backup := make([]int, 100)
		copy(backup, backends)
		subset := Subset(backup, clientID, subsetSize)
		fmt.Printf("subset: %v\n", subset)
	}
}

File: Week01/subset.py
import random


def Subset(backends, client_id, subset_size):
    subset_count = len(backends) // subset_size
    round = client_id // subset_count
    random.seed(round)
    random.shuffle(backends)
    subset_id = client_id % subset_count
    start = subset_id * subset_size
    return backends[start:start + subset_size]


if __name__ == '__main__':
    # client_id 范围 1 ~ 100000
    client_id = random.randint(1, 100000)
    # 100台backends 编号范围在0~10000之间 不连续 未排序
    backends = random.sample(range(10000), 100)
    # client 只和10台后端连接
    subset_size = 10
    for _ in range(10):
        # 传入为backends的拷贝
        subset = Subset(backends[:], client_id, subset_size)
        print(subset)
